### PR TITLE
Adyen add void, rebuild confirm, lock payment changes 

### PR DESCRIPTION
**Pull request by korycins**
- add cancel(void) to adyen
- rebuild confirm action to properly handle the async data
- add lock to confirm that we don't process the same transaction twice. (Thread race - adyen webhook vs checkoutComplete)

### CODE REVIEW

File: saleor/payment/gateway.py [hunk 1]
```diff
@@ -54,9 +57,20 @@ def wrapped(payment: Payment, *args, **kwargs):
     return wrapped
 
 
-@payment_postprocess
+def transaction_block(fn: Callable) -> Callable:
```
Comment (fowczarek) [inline saleor/payment/gateway.py:60]: The name is not clear from my perspective but I can't find better name. Maby we should add docstring for this decorator?

File: saleor/payment/gateway.py [hunk 2]
```diff
@@ -128,6 +154,10 @@ def capture(
     )
     if response and response.payment_method_info:
         update_payment_method_details(payment, response)
+    if response and response.transaction_already_processed:
+        transaction = get_already_processed_transaction(payment, response)
+        if transaction:
+            return transaction
```
Comment (fowczarek) [inline saleor/payment/gateway.py:160]: This code is duplicated in many places. Maby we should extract it to some utils?

File: saleor/payment/gateway.py (hunk 1 above)
Comment (maarcingebala) [reply saleor/payment/gateway.py:60]: Maybe `with_locked_payment`?

Comment (maarcingebala) [reply saleor/payment/gateway.py:60]: Plus a docstring.

File: saleor/payment/gateways/adyen/plugin.py [hunk 3]
```diff
@@ -442,3 +479,37 @@ def capture_payment(
...
+            Transaction.objects.filter(
+                payment__id=payment_information.payment_id,
+                kind=TransactionKind.AUTH,
+                is_success=True,
+            )
+            .exclude(token__isnull=True, token__exact="")
+            .last()
+        )
+        if not transaction:
+            raise PaymentError("Cannot find a payment reference to void.")
```
Comment (maarcingebala) [inline saleor/payment/gateways/adyen/plugin.py:497]: Just wondering - is this Adyen-specific or it should be part of `can_void`?

Comment (korycins) [reply saleor/payment/gateways/adyen/plugin.py:497]: It is specific for Adyen as any operation requires the reference received from the Auth operation.

**Pull request merged**